hands-on.ipynb

```diff
@@ -0,0 +1,499 @@
+    "data[\"load_balance\"].mean()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "cf2645e0-9b9a-4307-be5d-6573ae43ec4d",
+   "metadata": {},
+   "source": [
+    "On the author's machine this prints `0.25`, so with $N=4$ we are getting the worst possible load balance.\n",
+    "\n",
+    "**Can you see why?**"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "2f034260-e085-4558-bfd5-0a4151ea18b9",
+   "metadata": {},
+   "source": [
+    "### Single cell simulation\n",
+    "\n",
+    "In this example we are running `ringtest.py` with a single cell (`nring=1` and `ncell=1`) that is quite complex (`branch=5000` and `compart=5`):"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "70ea5657-93c6-4fc8-adcb-0bb554c21017",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "single_cell_args = (\n",
+    "    \"-nring\",\n",
+    "    1,\n",
+    "    \"-ncell\",\n",
+    "    1,\n",
+    "    \"-npt\",\n",
+    "    1,\n",
+    "    \"-branch\",\n",
+    "    5000,\n",
+    "    5000,\n",
+    "    \"-compart\",\n",
+    "    5,\n",
+    "    5,\n",
+    "    \"-tstop\",\n",
+    "    5,\n",
+    ")\n",
+    "single_cell_thread_data = {\n",
+    "    \"data\": {nt: ringtest(\"-nt\", nt, *single_cell_args) for nt in pows_of_2(max=8)},\n",
+    "    \"label\": \"Thread parallelism\",\n",
+    "}"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "eda78925-1286-47f6-8975-147ee8c66053",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "scaling_plot([single_cell_thread_data])"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "96d7d26c-9963-4704-b73d-894e809542dd",
+   "metadata": {},
+   "source": [
+    "Does this model benefit from using multiple threads? **Do you know why?**\n",
+    "\n",
+    "**What NEURON feature could we use to take advantage of multiple threads in this model?**"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "25b86856-75b9-4a41-80fb-81b87f2a692a",
+   "metadata": {},
+   "source": [
+    "*Next part is a spoiler/solution*"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "d12a3e56-99d5-40ca-b099-8ca61e950aac",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "single_cell_multisplit_thread_data = {\n",
+    "    \"data\": {\n",
+    "        nt: ringtest(\"-nt\", nt, \"-multisplit\", *single_cell_args)\n",
+    "        for nt in pows_of_2(max=8)\n",
+    "    },\n",
+    "    \"label\": \"Multisplit + threads\",\n",
+    "}"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "550526dd-ee3b-4ac2-98eb-66c4c4972a8b",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "scaling_plot([single_cell_thread_data, single_cell_multisplit_thread_data])"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "902b7615-0098-4f59-8dd9-627223aa6a17",
+   "metadata": {},
```

Comment:
@olupton : Just mentioning another thing that is in my mind considering some (important/historical) aspects/discussions about neuron performance (esp. [Table III](https://arxiv.org/pdf/1901.07454.pdf) in the 1st Arbor paper): how to get benefit from cache efficiency aspects (not neuron's cache efficiency option but benefit from CPU caches). 

As an example: consider there are 10 big cells in a ring and connected with [a larger delay](https://github.com/neuronsimulator/ringtest/blob/abb5b099966c71094b0250e2054a4af654b14b38/ring.py#L16) and with a typical sim timestep (dt) of `0.025`.

Assume these 10 threads together don't fit into a CPU cache. By default there will be 1 NrnThread and 10 cells will be computed for every single timestep.

One can set `pc.nthread(10, 0)` to split 10 cells into 10 different `NrnThread`s. The [2nd argument](https://nrn.readthedocs.io/en/latest/python/modelspec/programmatic/network/parcon.html#ParallelContext.nthread) `0` means internally there is a single Pthread (aka serial threads). With this, each NrnThread will be computed until the next communication interval (in this case 1msec / 0.025 msec = 40 time steps) i.e. each NrnThread object will be reused for 40 times.

Not saying we should demonstrate this with ringtest or include this! But just mentioning this topic as it has appeared in the past when people want to compare simulators without knowing this aspect. So putting here this blob FYI.

Reply:
👍